RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc

```diff
@@ -0,0 +1,198 @@
+    pfrh.depth() = HCAL::getDepth(pfrh.detId());
+    const uint32_t subdet = getSubdet(pfrh.detId());
+    if (subdet == HcalBarrel)
+      pfrh.layer() = PFLayer::HCAL_BARREL1;
+    else if (subdet == HcalEndcap)
+      pfrh.layer() = PFLayer::HCAL_ENDCAP;
+    else
+      pfrh.layer() = PFLayer::NONE;
+  }
+
+  template <>
+  ALPAKA_FN_ACC void PFRecHitProducerKernelConstruct<ECAL>::constructPFRecHit(
+      reco::PFRecHitDeviceCollection::View::element pfrh,
+      const ECAL::CaloRecHitSoATypeDevice::ConstView::const_element rh) {
+    pfrh.detId() = rh.detId();
+    pfrh.energy() = rh.energy();
+    pfrh.time() = rh.time();
+    pfrh.depth() = 1;
+    const uint32_t subdet = getSubdet(pfrh.detId());
+    if (subdet == EcalBarrel)
+      pfrh.layer() = PFLayer::ECAL_BARREL;
+    else if (subdet == EcalEndcap)
+      pfrh.layer() = PFLayer::ECAL_ENDCAP;
+    else
+      pfrh.layer() = PFLayer::NONE;
+  }
+
+  // Kernel to associate topology information of PFRecHits
+  template <typename CAL>
+  struct PFRecHitProducerKernelTopology {
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const typename CAL::TopologyTypeDevice::ConstView topology,
+                                  reco::PFRecHitDeviceCollection::View pfRecHits,
+                                  const uint32_t* __restrict__ denseId2pfRecHit,
+                                  uint32_t* __restrict__ num_pfRecHits) const {
+      // First thread updates size field pfRecHits SoA
+      if (const int32_t thread = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u]; thread == 0)
+        pfRecHits.size() = *num_pfRecHits;
+
+      // Assign position information and associate neighbours
+      for (int32_t i : cms::alpakatools::elements_with_stride(acc, *num_pfRecHits)) {
+        const uint32_t denseId = CAL::detId2denseId(pfRecHits.detId(i));
+
+        pfRecHits.x(i) = topology.positionX(denseId);
+        pfRecHits.y(i) = topology.positionY(denseId);
+        pfRecHits.z(i) = topology.positionZ(denseId);
+
+        for (uint32_t n = 0; n < 8; n++) {
+          pfRecHits.neighbours(i)(n) = -1;
+          const uint32_t denseId_neighbour = topology.neighbours(denseId)(n);
+          if (denseId_neighbour != 0xffffffff) {
+            const uint32_t pfRecHit_neighbour = denseId2pfRecHit[denseId_neighbour];
+            if (pfRecHit_neighbour != 0xffffffff)
+              pfRecHits.neighbours(i)(n) = (int32_t)pfRecHit_neighbour;
+          }
+        }
+      }
+    }
+  };
+
+  template <typename CAL>
+  PFRecHitProducerKernel<CAL>::PFRecHitProducerKernel(Queue& queue, const uint32_t num_recHits)
+      : denseId2pfRecHit_(cms::alpakatools::make_device_buffer<uint32_t[]>(queue, CAL::kSize)),
+        num_pfRecHits_(cms::alpakatools::make_device_buffer<uint32_t>(queue)),
+        work_div_(cms::alpakatools::make_workdiv<Acc1D>(1, 1)) {
```

Comment:
Is this explicit initialization needed? The actual initialization could be done here too e.g. along
```suggestion
        work_div_([](const uint32_t num_recHits) {
          const uint32_t items = 64;
          const uint32_t groups = cms::alpakatools::divide_up_by(num_recHits, items);
          return cms::alpakatools::make_workdiv<Acc1D>(groups, items);
        }(num_recHits)) {
```